core/domain/exp_services.py

```diff
@@ -642,33 +641,45 @@ def apply_change_list(
                 ].mark_written_translations_as_needing_update(change.content_id)
             elif change.cmd == exp_domain.CMD_EDIT_EXPLORATION_PROPERTY:
                 if change.property_name == 'title':
-                    exploration.update_title(change.new_value)  # type: ignore[no-untyped-call]
+                    exploration.update_title(change.new_value)
                 elif change.property_name == 'category':
-                    exploration.update_category(change.new_value)  # type: ignore[no-untyped-call]
+                    exploration.update_category(change.new_value)
                 elif change.property_name == 'objective':
-                    exploration.update_objective(change.new_value)  # type: ignore[no-untyped-call]
+                    exploration.update_objective(change.new_value)
                 elif change.property_name == 'language_code':
-                    exploration.update_language_code(change.new_value)  # type: ignore[no-untyped-call]
+                    exploration.update_language_code(change.new_value)
                 elif change.property_name == 'tags':
-                    exploration.update_tags(change.new_value)  # type: ignore[no-untyped-call]
+                    # Ruling out the possibility of any other type for mypy
+                    # type checking.
+                    assert isinstance(change.new_value, list)
```

Comment:
Not sure about this. Shouldn't this be checked in the controller layer?

As usual, any assertion needs a clear justification of why the condition could not possibly occur in practice, even if some other function written by a new developer in the future calls this one.

Reply:
Here, change is of type ExplorationChange which is a sub-class of BaseChange.

Every attribute ( other than cmd ) is created dynamically on instances of BaseChange class.
https://github.com/oppia/oppia/blob/60f19a72f06d50363572e318cf7d6c56e820dc34/core/domain/change_domain.py#L172

Also, here we defined the `__getattr__` method with return type str.

https://github.com/oppia/oppia/blob/60f19a72f06d50363572e318cf7d6c56e820dc34/core/domain/change_domain.py#L283

so every attribute which is created dynamically has type str. I did not altered the return type of `__getattr__` method because in codebase most of the attributes has str type. if we put union here then we have to use assertion in every place.

Reply:
Ditto for all places where we use assert to narrow down the type.

Reply:
As discussed in chat, I think the right thing to do here is to not use `__getattr__` for dynamic creation. Instead, we should make different named types or classes similar to what is done in the frontend for the different cases, and use the correct type/class for each type of change. See extensions/interactions/customization-args-defs.ts

Reply:
Encountered some issues, working on them -- till then can you please take look at other comments?

Reply:
Firstly, TypedDict is not going to work here because here we are accessing values as attributes, not keys.

Now, If we define a new domain object for each changeCmd, then we have to define all attributes of that changeCmd as a class variable to tell the mypy type of each attribute ( we cannot use __init__ here because we are accepting values as dictionary ). -- following example will explain things even more clearly.

```python
class TopicChange(change_domain.BaseChange):
    # Other things of topic_change
    pass ...

class DeleteSubtopicChangeCmd(TopicChange):    # this class represents `CMD_DELETE_SUBTOPIC` command.
    subtopic_id: str          # This is the required attribute name, which is used while making `CMD_DELETE_SUBTOPIC` command.

class AddCanonicalStoryChangeCmd(TopicChange):
    story_id: str              # This is the required attribute name, which is used while making `CMD_ADD_CANONICAL_STORY` command.

class UpdateSubtopicPropertyChangeCmd(TopicChange):
    subtopic_id: str           # This is the required attribute names, which is used while making `CMD_UPDATE_SUBTOPIC_PROPERTY` command.
    property_name: str
    new_value: Union[ union of all subtopic property types ]
    old_value: Union[ union of all subtopic property types ]


.
..
.
..
similarly all other commands.

```
Now, Currently we are using assert only when we need some other type from str (whereas str is the majority type -- I mean most of the change commands have str type ) but after implementing the above feature  ( or similar ) we have to use assert every time because in that case ( taking an example of above lines --  mentioned in this thread ) `change` is a union of all Change command domain classes ( Union[DeleteSubtopicChangeCmd, AddCanonicalStoryChangeCmd, UpdateSubtopicPropertyChangeCmd, ...]) and to use the `change` we have to first bring it down to the correct type using assert otherwise it will make the more confusion and throw more errors.


This is the most feasible approach that I get ( Keeping in mind the available info that I have ) -- please take a look and let me know if you have any other approach or if I missed something!!

Reply:
@seanlip @aasiffaizal PTAL!